middle_end/flambda2/to_cmm/to_cmm_env.ml

```diff
@@ -208,150 +323,407 @@ let get_exn_extra_args env k =
-     this involves moving the arguments, so they must be pure (or at most have
-     generative effects, with no coeffects). *)
-  match (effs : Ece.t), (extra : extra_info option) with
-  | ((No_effects | Only_generative_effects _), No_coeffects), Some Boxed_number
+let next_order = ref (-1)
+
+let simple cmm_expr = Simple { cmm_expr }
+
+let complex_no_split cmm_expr = Split { cmm_expr }
+
+let splittable_primitive dbg prim args = Splittable_prim { dbg; prim; args }
+
+let is_cmm_simple cmm =
+  match (cmm : Cmm.expression) with
+  | Cconst_int _ | Cconst_natint _ | Cconst_float _ | Cconst_symbol _ | Cvar _
     ->
     true
-  | ( ( (No_effects | Only_generative_effects _ | Arbitrary_effects),
-        (No_coeffects | Has_coeffects) ),
-      (None | Some Boxed_number | Some (Untag _)) ) ->
+  | Clet _ | Clet_mut _ | Cphantom_let _ | Cassign _ | Ctuple _ | Cop _
+  | Csequence _ | Cifthenelse _ | Cswitch _ | Ccatch _ | Cexit _ | Ctrywith _
+  | Cregion _ | Ctail _ ->
     false
 
-let create_binding =
-  let next_order = ref (-1) in
-  fun ?extra env ~may_inline effs var cmm_expr ->
-    let order =
-      incr next_order;
-      !next_order
+let create_binding_aux (type a) ?extra env effs var ~(inline : a inline)
+    (bound_expr : a bound_expr) =
+  let order =
+    let incr =
+      match bound_expr with
+      | Simple _ | Split _ -> 1
+      | Splittable_prim { args; _ } -> List.length args + 1
     in
-    let cmm_var = gen_variable var in
-    let binding = { order; may_inline; effs; cmm_var; cmm_expr } in
-    let cmm_expr = C.var (Backend_var.With_provenance.var cmm_var) in
-    let env = { env with vars = Variable.Map.add var cmm_expr env.vars } in
-    let env =
-      match extra with
-      | None -> env
-      | Some info ->
-        { env with vars_extra = Variable.Map.add var info env.vars_extra }
+    next_order := !next_order + incr;
+    !next_order
+  in
+  let cmm_var = gen_variable var in
+  let binding = Binding { order; inline; effs; cmm_var; bound_expr } in
+  let bindings = Variable.Map.add var binding env.bindings in
+  let cmm_expr = C.var (Backend_var.With_provenance.var cmm_var) in
+  let vars = Variable.Map.add var cmm_expr env.vars in
+  let vars_extra =
+    match extra with
+    | None -> env.vars_extra
+    | Some info -> Variable.Map.add var info env.vars_extra
+  in
+  (* Format.eprintf "*** new binding@\n%a@\n@." print_any_binding binding; *)
+  let env = { env with bindings; vars; vars_extra } in
+  env, binding
+
+let create_binding (type a) ?extra env effs var ~(inline : a inline)
+    (bound_expr : a bound_expr) =
+  (* In order to avoid generating binding of the form: "let x = y in ...", when
+     'y' is trivial i.e. is a value that fits in a register, we mark 'x' as a
+     must_inline_and_duplicate (since it basically replaces a variable by either
+     another variable, a constant, or a symbol). *)
+  match bound_expr with
+  | Simple { cmm_expr } when is_cmm_simple cmm_expr ->
```

Comment:
This could also apply on `Split` bindings to a simple expression.

Reply:
done